### PR TITLE
Homework 7 (JDI)

### PR DESCRIPTION
**Pull request by Katsiaryna-Ledak**

### CODE REVIEW

File: src/test/resources/HW3.properties [hunk 1]
```diff
@@ -2,3 +2,5 @@
 userName = Roman
```
Comment (pnatashap) [inline src/test/resources/HW3.properties:2]: Should not be a part of this PR

File: src/test/resources/testng/ex1Ex2HW4.xml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+<!DOCTYPE suite SYSTEM "https://testng.org/testng-1.0.dtd">
+
```
Comment (pnatashap) [inline src/test/resources/testng/ex1Ex2HW4.xml:2]: Should not be a part of this PR

File: src/test/resources/testng/ex1Ex2HW6.xml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+<!DOCTYPE suite SYSTEM "https://testng.org/testng-1.0.dtd">
+
+<suite name="Selenium Grid Suite">
+  <test name="ex1Ex2HW6Test">
```
Comment (pnatashap) [inline src/test/resources/testng/ex1Ex2HW6.xml:4]: Should not be a part of this PR

File: src/test/java/com/epam/tc/hw7/tests/BaseTest.java [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    @BeforeSuite(alwaysRun = true)
+    static void setUp() {
+        killAllSeleniumDrivers();
+        openSite(JdiTestingSite.class);
+
+        // Login on JDI site as User
+        homePage.open();
+        userIcon.click();
+        loginForm.loginAs(DEFAULT_USER);
+        homePage.checkOpened();
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/tests/BaseTest.java:25]: Test is failed on MaterialAndMetals page open if login is failed. 
We should fail on login

**Review by pnatashap — CHANGES_REQUESTED** [review]
Comments in code

File: src/test/java/com/epam/tc/hw3/pages/HomePage.java [hunk 5]
```diff
@@ -85,4 +85,5 @@ public void clickOnServiceDiffElementsHead() {
         headMenuElement.get(2).click();
         getServiceDiffElementsHeadMenu.click();
     }
+
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw3/pages/HomePage.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/pages/AbstractPage.java [hunk 6]
```diff
@@ -0,0 +1,12 @@
+package com.epam.tc.hw4.pages;
+
+import org.openqa.selenium.WebDriver;
+
+public class AbstractPage {
+
+    protected WebDriver webDriver;
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/pages/AbstractPage.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/pages/DifferentElementsPage.java [hunk 7]
```diff
@@ -0,0 +1,59 @@
+package com.epam.tc.hw4.pages;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+public class DifferentElementsPage extends AbstractPage {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/pages/DifferentElementsPage.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/pages/HomePage.java [hunk 8]
```diff
@@ -0,0 +1,91 @@
+package com.epam.tc.hw4.pages;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+public class HomePage extends AbstractPage {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/pages/HomePage.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/pages/LoginPanelComponent.java [hunk 9]
```diff
@@ -0,0 +1,39 @@
+package com.epam.tc.hw4.pages;
+
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+public class LoginPanelComponent extends AbstractPage {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/pages/LoginPanelComponent.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/steps/Exercise1Steps.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+import com.epam.tc.hw4.pages.HomePage;
+import com.epam.tc.hw4.pages.LoginPanelComponent;
+import io.qameta.allure.Step;
+import java.util.List;
+import org.assertj.core.api.Assertions;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+
+
+public class Exercise1Steps {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/steps/Exercise1Steps.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/steps/Exersice2Steps.java [hunk 11]
```diff
@@ -0,0 +1,75 @@
...
+
+import com.epam.tc.hw4.pages.DifferentElementsPage;
+import com.epam.tc.hw4.pages.HomePage;
+import com.epam.tc.hw4.pages.LoginPanelComponent;
+import io.qameta.allure.Step;
+import java.util.List;
+import org.assertj.core.api.Assertions;
+import org.openqa.selenium.WebDriver;
+
+public class Exersice2Steps {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/steps/Exersice2Steps.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/testcases/DifferentElementsPageTC.java [hunk 12]
```diff
@@ -0,0 +1,57 @@
+package com.epam.tc.hw4.testcases;
+
+import com.epam.tc.hw4.steps.Exersice2Steps;
+import com.epam.tc.hw4.util.PropertiesFileReader;
+import io.qameta.allure.Feature;
+import io.qameta.allure.Story;
+import org.testng.annotations.Test;
+
+public class DifferentElementsPageTC extends InitializationTest {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/testcases/DifferentElementsPageTC.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/testcases/HomePageTC.java [hunk 13]
```diff
@@ -0,0 +1,52 @@
+package com.epam.tc.hw4.testcases;
+
+import com.epam.tc.hw4.steps.Exercise1Steps;
+import com.epam.tc.hw4.util.PropertiesFileReader;
+import io.qameta.allure.Feature;
+import io.qameta.allure.Story;
+import org.testng.annotations.Test;
+
+public class HomePageTC extends InitializationTest {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/testcases/HomePageTC.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/testcases/InitializationTest.java [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.chrome.ChromeDriver;
+import org.testng.ITestContext;
+import org.testng.annotations.AfterMethod;
+import org.testng.annotations.BeforeMethod;
+
+public class InitializationTest {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/testcases/InitializationTest.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/util/AttachmentUtils.java [hunk 15]
```diff
@@ -0,0 +1,11 @@
+package com.epam.tc.hw4.util;
+
+import io.qameta.allure.Attachment;
+
+public class AttachmentUtils {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/util/AttachmentUtils.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/util/PropertiesFileReader.java [hunk 16]
```diff
@@ -0,0 +1,24 @@
+package com.epam.tc.hw4.util;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+public class PropertiesFileReader {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/util/PropertiesFileReader.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/pages/DifferentElementsPage.java [hunk 17]
```diff
@@ -0,0 +1,58 @@
+package com.epam.tc.hw6.pages;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+public class DifferentElementsPage extends AbstractPage {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/pages/DifferentElementsPage.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/pages/HomePage.java [hunk 18]
```diff
@@ -0,0 +1,91 @@
+package com.epam.tc.hw6.pages;
+
+import java.util.ArrayList;
+import java.util.List;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+public class HomePage extends AbstractPage {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/pages/HomePage.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/pages/LoginPanelComponent.java [hunk 19]
```diff
@@ -0,0 +1,39 @@
+package com.epam.tc.hw6.pages;
+
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+import org.openqa.selenium.support.PageFactory;
+
+public class LoginPanelComponent extends AbstractPage {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/pages/LoginPanelComponent.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/steps/Exercise1Steps.java [hunk 20]
```diff
@@ -0,0 +1,119 @@
...
+
+import com.epam.tc.hw6.pages.HomePage;
+import com.epam.tc.hw6.pages.LoginPanelComponent;
+import io.qameta.allure.Step;
+import java.util.List;
+import org.assertj.core.api.Assertions;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+
+public class Exercise1Steps {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/steps/Exercise1Steps.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/steps/Exersice2Steps.java [hunk 21]
```diff
@@ -0,0 +1,75 @@
...
+
+import com.epam.tc.hw6.pages.DifferentElementsPage;
+import com.epam.tc.hw6.pages.HomePage;
+import com.epam.tc.hw6.pages.LoginPanelComponent;
+import io.qameta.allure.Step;
+import java.util.List;
+import org.assertj.core.api.Assertions;
+import org.openqa.selenium.WebDriver;
+
+public class Exersice2Steps {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/steps/Exersice2Steps.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/testcases/DifferentElementsPageTC.java [hunk 22]
```diff
@@ -0,0 +1,57 @@
+package com.epam.tc.hw6.testcases;
+
+import com.epam.tc.hw6.steps.Exersice2Steps;
+import com.epam.tc.hw6.util.PropertiesFileReader;
+import io.qameta.allure.Feature;
+import io.qameta.allure.Story;
+import org.testng.annotations.Test;
+
+public class DifferentElementsPageTC extends InitializationTest {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/testcases/DifferentElementsPageTC.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/testcases/HomePageTC.java [hunk 23]
```diff
@@ -0,0 +1,52 @@
+package com.epam.tc.hw6.testcases;
+
+import com.epam.tc.hw6.steps.Exercise1Steps;
+import com.epam.tc.hw6.util.PropertiesFileReader;
+import io.qameta.allure.Feature;
+import io.qameta.allure.Story;
+import org.testng.annotations.Test;
+
+public class HomePageTC extends InitializationTest {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/testcases/HomePageTC.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/util/AttachmentUtils.java [hunk 24]
```diff
@@ -0,0 +1,11 @@
+package com.epam.tc.hw6.util;
+
+import io.qameta.allure.Attachment;
+
+public class AttachmentUtils {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/util/AttachmentUtils.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/util/PropertiesFileReader.java [hunk 25]
```diff
@@ -0,0 +1,24 @@
+package com.epam.tc.hw6.util;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Properties;
+
+public class PropertiesFileReader {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/util/PropertiesFileReader.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/util/driver/WebDriverFactory.java [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+import org.openqa.selenium.Capabilities;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.chrome.ChromeDriver;
+import org.openqa.selenium.chrome.ChromeOptions;
+import org.openqa.selenium.firefox.FirefoxDriver;
+import org.openqa.selenium.firefox.FirefoxOptions;
+import org.openqa.selenium.remote.RemoteWebDriver;
+
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class WebDriverFactory {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/util/driver/WebDriverFactory.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/util/driver/WebDriverSingleton.java [hunk 27]
```diff
@@ -0,0 +1,23 @@
+package com.epam.tc.hw6.util.driver;
+
+import java.util.Objects;
+import org.openqa.selenium.WebDriver;
+
+public class WebDriverSingleton {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/util/driver/WebDriverSingleton.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw6/util/driver/WebDriverTypeException.java [hunk 28]
```diff
@@ -0,0 +1,8 @@
+package com.epam.tc.hw6.util.driver;
+
+public class WebDriverTypeException extends RuntimeException {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw6/util/driver/WebDriverTypeException.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/resources/testng/ex1Ex2HW4.xml (hunk 2 above)
Comment (Katsiaryna-Ledak) [reply src/test/resources/testng/ex1Ex2HW4.xml]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw4/listeners/ScreenshotListener.java [hunk 29]
```diff
@@ -0,0 +1,20 @@
+package com.epam.tc.hw4.listeners;
+
+import com.epam.tc.hw4.util.AttachmentUtils;
+import org.openqa.selenium.OutputType;
+import org.openqa.selenium.TakesScreenshot;
+import org.testng.ITestListener;
+import org.testng.ITestResult;
+
+public class ScreenshotListener implements ITestListener {
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw4/listeners/ScreenshotListener.java]: After rebase with master this comment is marked as 'outdated'

File: src/test/java/com/epam/tc/hw7/tests/BaseTest.java (hunk 4 above)
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw7/tests/BaseTest.java:29]: Resolved in commit:https://github.com/Katsiaryna-Ledak/Ekaterina_Khudzhamkulova/pull/8/commits/e01eb318d1b38ea5a4432e348a944597702841da

File: src/test/java/com/epam/tc/hw7/tests/BaseTest.java [hunk 30]
```diff
@@ -0,0 +1,32 @@
...
+import com.epam.tc.hw7.site.JdiTestingSite;
+import org.testng.annotations.AfterSuite;
+import org.testng.annotations.BeforeSuite;
+
+public class BaseTest {
+
+    @BeforeSuite(alwaysRun = true)
+    static void setUp() {
+        killAllSeleniumDrivers();
+        openSite(JdiTestingSite.class);
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw7/tests/BaseTest.java]: Resolved in commit:https://github.com/Katsiaryna-Ledak/Ekaterina_Khudzhamkulova/pull/8/commits/e01eb318d1b38ea5a4432e348a944597702841da

File: src/test/java/com/epam/tc/hw7/dataproviders/UserData.java [hunk 31]
```diff
@@ -0,0 +1,12 @@
+package com.epam.tc.hw7.dataproviders;
+
+import com.epam.tc.hw7.entities.User;
+
+public class UserData {
+
+    public static User DEFAULT_USER = new User().set(
+        u -> {
+            u.userName = "Roman";
+            u.password = "Jdi1234";
```
[the remark this answers is not in the document]
Comment (Katsiaryna-Ledak) [reply src/test/java/com/epam/tc/hw7/dataproviders/UserData.java:11]: Resolved in commit:https://github.com/Katsiaryna-Ledak/Ekaterina_Khudzhamkulova/pull/8/commits/e01eb318d1b38ea5a4432e348a944597702841da

File: src/test/java/com/epam/tc/hw7/tests/BaseTest.java [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+        loginForm.loginAs(DEFAULT_USER);
+
+        // Assert if Login in correct
+        userName.is().text(DEFAULT_USER.fullName);
+        homePage.checkOpened();
+    }
+
+    @AfterSuite(alwaysRun = true)
+    static void teardown() {
+        homePage.logout();
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/tests/BaseTest.java:34]: What is the reason to logout if we kill the browser after?

File: src/test/java/com/epam/tc/hw7/dataproviders/MetalsColorsDataProvider.java [hunk 33]
```diff
@@ -0,0 +1,22 @@
...
+
+public class MetalsColorsDataProvider {
+
+    @DataProvider(name = "metalsColorsObjects")
+    public static Object[][] metalsColorsObjects() throws IOException {
+
+        InputStream fileInputStream = MetalsColorsDataProvider
+            .class.getResourceAsStream("/json/JDI_ex8_metalsColorsDataSet.json");
+
+        String jsonText = IOUtils.toString(fileInputStream, StandardCharsets.UTF_8.name());
```
Comment (pnatashap) [inline src/test/java/com/epam/tc/hw7/dataproviders/MetalsColorsDataProvider.java:18]: It's not a good practice to read content, Gson can be used with Reader.

**Review by pnatashap — APPROVED** [review]
There are two comments, it's better to check the best practices for these situations

**Pull request merged**